### PR TITLE
File serve from multi adapter 

### PR DESCRIPTION
**Pull request by kasvith**
### Summary of Changes
- Added file serving method to API
- Added file serving invoking to Controller

### Testing Instructions
- Log in to backend
- Navigate to following URL 

http://localhost/media-manager-improvement/administrator/index.php?option=com_media&format=json&url=true&task=api.files&path=local:/banners/banner.jpg

### Expected result
- AJAX response with data containing a url to banner.jpg

### CODE REVIEW

**Comment by kasvith** [comment]
Tarvis will fail as i ve changed ctr a bit
Once code is acceptable i will fix them and we can  merge then

**Comment by laoneo** [comment]
I'v merged the milestone1 branch into the gsoc17. Please merge now the gsoc17 branch into yours. Then the commit log should show up fine. Hopefully.

**Comment by kasvith** [comment]
@laoneo done, can you check it?

**Comment by kasvith** [comment]
tarvis will fail as i have not updated test cases

File: .gitignore [hunk 1]
```diff
@@ -209,6 +209,7 @@ Desktop.ini
 /libraries/vendor/zendframework/zend-diactoros/CONTRIBUTING.md
 /libraries/vendor/zendframework/zend-diactoros/mkdocs.yml
 /libraries/vendor/zendframework/zend-diactoros/README.md
+/plugins/filesystem/dropbox/composer.lock
```
Comment (laoneo) [inline .gitignore]: Please remove that line.

File: administrator/components/com_media/Adapter/AdapterInterface.php [hunk 2]
```diff
@@ -149,4 +149,17 @@ public function move($sourcePath, $destinationPath, $force = false);
 	 * @throws  \Exception
 	 */
 	public function copy($sourcePath, $destinationPath, $force = false);
+
```
Comment (laoneo) [inline administrator/components/com_media/Adapter/AdapterInterface.php:152]: One empty line to much.

File: administrator/components/com_media/Adapter/AdapterInterface.php [hunk 3]
```diff
@@ -149,4 +149,17 @@ public function move($sourcePath, $destinationPath, $force = false);
...
+	 * Returns a permanent link for media file.
+	 *
+	 * @param   string  $path  The path to file
+	 *
+	 * @return string
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 * @throws FileNotFoundException
+	 */
+	public function getPermalink($path);
```
Comment (laoneo) [inline administrator/components/com_media/Adapter/AdapterInterface.php]: As discussed I would change that to getUrl or something more meaningful.

File: administrator/components/com_media/Controller/Api.php [hunk 4]
```diff
@@ -151,7 +165,7 @@ public function files()
 			// Return the data
 			$this->sendResponse($data);
 		}
-		catch (MediaFileAdapterFilenotfoundexception $e)
+		catch (FileNotFoundException $e)
```
Comment (laoneo) [inline administrator/components/com_media/Controller/Api.php]: Can you make a pr with this fix against the milestone1 branch? We need to have it fixed there as well.

File: administrator/components/com_media/Controller/Api.php [hunk 5]
```diff
@@ -104,7 +105,20 @@ public function files()
 			switch ($method)
 			{
 				case 'get':
-					$data = $this->getModel()->getFiles($adapter, $path, $this->input->getWord('filter'));
+					$option = $this->input->getWord('opt');
```
Comment (laoneo) [inline administrator/components/com_media/Controller/Api.php]: I would rather make it as an option that the get response should include the permalink or not in the returned data. IMO the GET response should always return the same data, but with some more details or not.

File: plugins/filesystem/local/Adapter/LocalAdapter.php [hunk 6]
```diff
@@ -544,4 +554,19 @@ private function moveFolder($sourcePath, $destinationPath, $force = false)
...
+	 *
+	 * @param   string  $path  Path of the file relative to adapter
+	 *
+	 * @return string
+	 *
+	 * @since __DEPLOY_VERSION__
+	 */
+	public function getPermalink($path)
+	{
+		return \Joomla\CMS\Uri\Uri::root() . \JPath::clean($this->filePath . $path);
```
Comment (laoneo) [inline plugins/filesystem/local/Adapter/LocalAdapter.php]: No need to add here the full namespace as you made the use statement already.

**Review by laoneo — CHANGES_REQUESTED** [review]

File: administrator/components/com_media/Controller/Api.php (hunk 4 above)
Comment (kasvith) [reply administrator/components/com_media/Controller/Api.php]: Sure

File: administrator/components/com_media/Controller/Api.php (hunk 5 above)
Comment (kasvith) [reply administrator/components/com_media/Controller/Api.php]: I did not get that, anyway do we want an assoc array?

File: administrator/components/com_media/Adapter/AdapterInterface.php (hunk 2 above)
Comment (kasvith) [reply administrator/components/com_media/Adapter/AdapterInterface.php:152]: it is updated removed line follows that

File: administrator/components/com_media/Controller/Api.php (hunk 5 above)
Comment (kasvith) [reply administrator/components/com_media/Controller/Api.php]: No actually as we discussed we only need to get URL iff requested, so IMO we should not send it along file attributes at all, only when it requested we send URL along with request for selected file

Comment (kasvith) [reply administrator/components/com_media/Controller/Api.php]: If you want we can create another task

Comment (kasvith) [reply administrator/components/com_media/Controller/Api.php]: @dneukirchen do we need something like url => "path/to/url" or just "path/to/url" along with data in AJAX response ?

File: administrator/components/com_media/Controller/Api.php [hunk 7]
```diff
@@ -104,7 +105,20 @@ public function files()
...
+						case 'permalink':
+							$data = $this->getModel()->getUrl($adapter, $path);
+							break;
+
+						default:
+							$data = $this->getModel()->getFiles($adapter, $path, $this->input->getWord('filter'));
+							break;
+
+					}
+
```
Comment (dgrammatiko) [inline administrator/components/com_media/Controller/Api.php]: @kasvith I think an if else is better here (`if (isset($option) && $option === 'permalink')`

File: administrator/components/com_media/Controller/Api.php (hunk 4 above)
Comment (kasvith) [reply administrator/components/com_media/Controller/Api.php]: i ve made pr

File: administrator/components/com_media/Controller/Api.php (hunk 7 above)
Comment (kasvith) [reply administrator/components/com_media/Controller/Api.php]: ya an if would do the job also

File: administrator/components/com_media/Controller/Api.php (hunk 5 above)
Comment (laoneo) [reply administrator/components/com_media/Controller/Api.php]: For example the google calendar api has a flag `alwaysIncludeEmail` which returns the email addresses in the response or not https://developers.google.com/google-apps/calendar/v3/reference/events/list. I would rather use something like that then returning a totally different response on a GET request for the same endpoint.

File: administrator/components/com_media/resources/scripts/store/state.js [hunk 8]
```diff
@@ -23,6 +23,4 @@ export default {
     selectedDirectory: options.providers[0].name + ':/',
     // The currently selected items
     selectedItems: [],
-    // The currently selected items
-    selectedItems: [],
```
Comment (laoneo) [inline administrator/components/com_media/resources/scripts/store/state.js]: Why is this file modified?

File: administrator/components/com_media/Adapter/AdapterInterface.php [hunk 9]
```diff
@@ -149,4 +149,16 @@ public function move($sourcePath, $destinationPath, $force = false);
 	 * @throws  \Exception
 	 */
 	public function copy($sourcePath, $destinationPath, $force = false);
+
+	/**
+	 * Returns a permanent link for media file.
```
Comment (laoneo) [inline administrator/components/com_media/Adapter/AdapterInterface.php]: Please add here a better comment what's the purpose of this function.

File: administrator/components/com_media/Controller/Api.php [hunk 10]
```diff
@@ -104,7 +105,11 @@ public function files()
 			switch ($method)
 			{
 				case 'get':
-					$data = $this->getModel()->getFiles($adapter, $path, $this->input->getWord('filter'));
+					$options = array();
+					$options[] = $this->input->getWord('opt');
```
Comment (laoneo) [inline administrator/components/com_media/Controller/Api.php]: I would make it an indexed array, where the key is here url and the value true or false.

File: administrator/components/com_media/Controller/Api.php (hunk 4 above)
Comment (laoneo) [reply administrator/components/com_media/Controller/Api.php]: Please update your branch with the gsoc17, merged your pr already.

File: administrator/components/com_media/Model/Api.php [hunk 11]
```diff
@@ -120,23 +121,30 @@ public function getFile($adapter, $path = '/')
...
+		// Add adapter prefix to all the files to be returned
 		$files = $this->getAdapter($adapter)->getFiles($path, $filter);
 
 		foreach ($files as $file)
 		{
 			$file->path = $adapter . ":" . $file->path;
 		}
 
+		// Add url info to a file when option is set
+		if (in_array("url", $options) && count($files) === 1 && $files[0]->type == "file")
```
Comment (laoneo) [inline administrator/components/com_media/Model/Api.php]: ` && count($files) === 1 && $files[0]->type == "file"` can be removed, if somebody wants to have it for all files, then it is in his responsibility. We should not force that. I'm aware that it can be a performance issue, but so it is.

**Review by laoneo — CHANGES_REQUESTED** [review]

File: administrator/components/com_media/Controller/Api.php (hunk 10 above)
Comment (kasvith) [reply administrator/components/com_media/Controller/Api.php]: ahh thats a good idea

File: administrator/components/com_media/resources/scripts/store/state.js (hunk 8 above)
Comment (kasvith) [inline administrator/components/com_media/resources/scripts/store/state.js]: tried some experiment to fix create folder model, will revert it

File: administrator/components/com_media/Model/Api.php (hunk 11 above)
Comment (kasvith) [inline administrator/components/com_media/Model/Api.php]: IMO method getUrl only should act on files, in here we are actually giving a folder path as the input path to getUrl and 3dp may be mislead to implement it for a whole folder.

If someone break this plugins will be malfunction IMO

File: administrator/components/com_media/resources/scripts/store/state.js (hunk 8 above)
Comment (kasvith) [reply administrator/components/com_media/resources/scripts/store/state.js]: there was a duplicated variable i removed it

File: administrator/components/com_media/Model/Api.php (hunk 11 above)
Comment (kasvith) [inline administrator/components/com_media/Model/Api.php]: even we added url for files in a directory if it contain a directory it makes no sense, IMO we should force them to apply this only on a single file. Since 3dps will never touch this we can make em force to avoid unnecessary checks in api

File: administrator/components/com_media/resources/scripts/store/state.js (hunk 8 above)
Comment (laoneo) [reply administrator/components/com_media/resources/scripts/store/state.js]: I would do that in another pr against the milestone 1 branch. One thing in one pr.

File: tests/unit/suites/plugins/filesystem/local/LocalAdapterTest.php [hunk 12]
```diff
@@ -26,6 +26,13 @@ class LocalAdapterTest extends TestCaseDatabase
 	private $root = null;
 
 	/**
+	 * The image folder path related to root
+	 * 
+	 * @var null
+	 */
+	private $image_path = null;
```
Comment (laoneo) [inline tests/unit/suites/plugins/filesystem/local/LocalAdapterTest.php]: Please camel case

**Review by laoneo — CHANGES_REQUESTED** [review]

**Comment by laoneo** [comment]
Would be good to also add a test for getUrl

File: administrator/components/com_media/resources/scripts/store/state.js (hunk 8 above)
Comment (kasvith) [reply administrator/components/com_media/resources/scripts/store/state.js]: Ahh sorry about that

File: administrator/components/com_media/Model/Api.php [hunk 13]
```diff
@@ -120,23 +121,32 @@ public function getFile($adapter, $path = '/')
...
 		foreach ($files as $file)
 		{
+			// If requested add options
+			if (isset($options['url']) && $options['url'] == 'true' && $file->type == 'file')
+			{
+				$file->url = $this->getUrl($adapter, $file->path);
+			}
+
 			$file->path = $adapter . ":" . $file->path;
+
```
Comment (laoneo) [inline administrator/components/com_media/Model/Api.php]: Empty line not needed

File: administrator/components/com_media/Model/Api.php [hunk 14]
```diff
@@ -120,23 +121,32 @@ public function getFile($adapter, $path = '/')
...
 		}
 
+
```
Comment (laoneo) [inline administrator/components/com_media/Model/Api.php]: Empty line not needed

**Review by laoneo — CHANGES_REQUESTED** [review]

**Comment by laoneo** [comment]
What did you do for changes in mediamanager.js?

**Comment by kasvith** [comment]
@laoneo i have run build-js again after changing the state.js

**Comment by laoneo** [comment]
But you reverted that change, so you have to revert that file then as well.

File: administrator/components/com_media/Model/Api.php [hunk 15]
```diff
@@ -120,20 +121,27 @@ public function getFile($adapter, $path = '/')
...
+	public function getFiles($adapter, $path = '/', $filter = '', $options = array())
 	{
-		// Add adapter info to all the files to be returned
+		// Add adapter prefix to all the files to be returned
 		$files = $this->getAdapter($adapter)->getFiles($path, $filter);
 
 		foreach ($files as $file)
 		{
+			// If requested add options
+			if (isset($options['url']) && $options['url'] == 'true' && $file->type == 'file')
```
Comment (laoneo) [inline administrator/components/com_media/Model/Api.php]: I would not test if the string is `'true'`. I would just do `if (isset($options['url']) && $options['url'] && $file->type == 'file')`.

File: administrator/components/com_media/Controller/Api.php [hunk 16]
```diff
@@ -105,7 +105,12 @@ public function files()
 			switch ($method)
 			{
 				case 'get':
-					$data = $this->getModel()->getFiles($adapter, $path, $this->input->getWord('filter'));
+					// Grab options
+					$options = array();
+					$options['url'] = $this->input->getWord('url', '');
```
Comment (laoneo) [inline administrator/components/com_media/Controller/Api.php]: I would cast that here to boolean, see my comment below why.

**Review by laoneo — CHANGES_REQUESTED** [review]

**Comment by laoneo** [comment]
Thank you very much!

**Comment by kasvith** [comment]
thanks @laoneo for spending time

**Comment by dneukirchen** [comment]
gj @kasvith
